workflows/src/main/kotlin/com/r3/corda/lib/ci/RequestKeyFlow.kt

```diff
@@ -0,0 +1,100 @@
+package com.r3.corda.lib.ci
+
+import co.paralleluniverse.fibers.Suspendable
+import net.corda.core.contracts.UniqueIdentifier
+import net.corda.core.crypto.SecureHash
+import net.corda.core.flows.FlowException
+import net.corda.core.flows.FlowLogic
+import net.corda.core.flows.FlowSession
+import net.corda.core.identity.Party
+import net.corda.core.serialization.deserialize
+import net.corda.core.utilities.ProgressTracker
+import net.corda.core.utilities.unwrap
+import java.security.PublicKey
+import java.util.*
+
+/**
+ * This flow registers a mapping in the [net.corda.core.node.services.IdentityService] between a [PublicKey] and a [Party]. It can generate a new key
+ * pair for a given [UUID] and register the new key mapping, or a known [PublicKey] can be supplied to the flow which will register
+ * a mapping between this key and the requesting party.
+ *
+ * The generation of the [SignedKeyForAccount] is delegated to the counter-party which concatenates the original [ChallengeResponse] with its own
+ * [ChallengeResponse] and signs over the concatenated hash before sending this value and the [PublicKey] and sends it back to the requesting node.
+ * The requesting node verifies the signature on the [ChallengeResponse] and verifies the concatenated [ChallengeResponse] is the same as the one received
+ * from the counter-party.
+ */
+class RequestKeyFlow
+private constructor(
+        private val session: FlowSession,
+        private val uuid: UUID?,
+        private val key: PublicKey?) : FlowLogic<SignedKeyForAccount>() {
+    constructor(session: FlowSession, uuid: UUID) : this(session, uuid, null)
+    constructor(session: FlowSession, key: PublicKey) : this(session, UniqueIdentifier().id, key)
+    constructor(session: FlowSession) : this(session, null, null)
+
+    companion object {
+        object REQUESTING_KEY : ProgressTracker.Step("Requesting a public key")
+        object VERIFYING_KEY : ProgressTracker.Step("Verifying counterparty's signature")
+        object KEY_VERIFIED : ProgressTracker.Step("Signature is correct")
+        object VERIFYING_CHALLENGE_RESPONSE : ProgressTracker.Step("Verifying the received SHA-256 matches the original that was sent")
+        object CHALLENGE_RESPONSE_VERIFIED : ProgressTracker.Step("SHA-256 is correct")
+
+        @JvmStatic
+        fun tracker(): ProgressTracker = ProgressTracker(REQUESTING_KEY, VERIFYING_KEY, KEY_VERIFIED, VERIFYING_CHALLENGE_RESPONSE, CHALLENGE_RESPONSE_VERIFIED)
+    }
+
+    override val progressTracker = tracker()
+
+    @Suspendable
+    @Throws(FlowException::class)
+    override fun call(): SignedKeyForAccount {
+        progressTracker.currentStep = REQUESTING_KEY
+        val challengeResponseParam = SecureHash.randomSHA256()
+        val requestKey = if (key == null && uuid != null ) {
```

Comment:
I always find it's easier to use `when` statements in situations like this e.g.

```
fun test(key: PublicKey?, uuid: UUID?) {
    val noKey = key == null
    val noUuid = uuid == null
    when {
        noKey && noUuid -> doThis()
        noKey && !noUuid -> doThat()
        !noKey && noUuid -> doOther()
    }
}
```

Reply:
Done